dvc/repo/ls.py

```diff
@@ -0,0 +1,108 @@
+import os
+
+from dvc.exceptions import DvcException, NotDvcRepoError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    if os.path.exists(url):
+        if rev:
+            raise DvcException("Local repo doesn't support revisions")
+        return _ls_local_repo(url, target, recursive, outputs_only)
+
+    return _ls_remote_git_repo(url, target, rev, recursive, outputs_only)
+
+
+def _ls_files_local_repo(path, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+
+    def prettify(prefix, filename):
+        if not prefix == os.curdir:
+            filename = os.path.join(prefix, filename)
+        return filename
+
+    files = []
+    target_path = os.path.join(path, target) if target else path
+    tree = CleanTree(WorkingTree(target_path))
+
+    try:
+        target_path = os.path.join(path, target) if target else path
+        if not os.path.exists(target_path):
+            raise FileNotFoundError("{} doesn't exist".format(target_path))
+
+        if target:
+            files.append(prettify(os.curdir, target))
+            if not recursive:
+                return files
+
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            prefix = os.path.relpath(dirpath, path)
+            files.extend(map(lambda f: prettify(prefix, f), dirnames))
+            files.extend(map(lambda f: prettify(prefix, f), filenames))
+            if not recursive:
+                break
+    except (NotADirectoryError, FileNotFoundError):
+        pass
+
+    return files
+
+
+def _ls_outs_local_repo(path, target=None, recursive=None):
+    from dvc.repo import Repo
+    from dvc.path_info import PathInfo
+
+    try:
+        repo = Repo(path)
+    except NotDvcRepoError:
+        return []
+
+    outs = [out for stage in repo.stages for out in stage.outs]
+    target_path = os.path.join(path, target) if target else path
+    path_info = PathInfo(os.path.abspath(target_path))
+
+    def prettify(out):
+        relpath = out.path_info.relpath(path)
+        parts = list(relpath.parts)
+        segments = [parts.pop(0)]
+        if not recursive:
+            return segments
+
+        while parts:
+            segments.append(os.path.join(segments[-1], parts.pop(0)))
+        return segments
+
+    files = set()
+    for out in outs:
```

Comment:
Since we no longer care about multiple targets, could we now use find_outs_by_path/relpath here?

Reply:
I don't like `repo.find_outs_by_path/relpath` cause:

## Case#1
I request `dvc list repo`  

`repo.find_outs_by_path` return noting on trying to get content of the repo without `recursive=True`, what requires always use it with `recursive` and filter then the output depends on the depth
```
>>> os.getcwd()
'/home/jiojiajiu/workspace/dvc/test-project'
>>> repo = Repo(os.getcwd())
>>> repo.find_outs_by_path('.')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jiojiajiu/workspace/dvc/dvc/repo/__init__.py", line 416, in find_outs_by_path
    raise OutputNotFoundError(path, self)
dvc.exceptions.OutputNotFoundError: Unable to find DVC-file with output '.'
>>> repo.find_outs_by_path('.', recursive=True)
[OutputLOCAL: 'dup/1', OutputLOCAL: '1', OutputLOCAL: 'add_dir', OutputLOCAL: 'model.pkl', OutputLOCAL: 'data/prepared', OutputLOCAL: 'test_dir', OutputLOCAL: 'auc.metric', OutputLOCAL: 'data/features', OutputLOCAL: '2', OutputLOCAL: 'data.xml']
```
but
```
(dvc-get-started) jiojiajiu@pon4ik:~/workspace/dvc/test-project$ dvc list . --outputs-only
1
2
add_dir
auc.metric
data
dup
model.pkl
test_dir
```

Reply:
Might be missing something, but isn't that what you currently do? You filter-out all of the outputs and then do some segments magic to get only first part of the path.

Reply:
I think you are right @efiop, my bad